File: test/csharp/SpecGate.CtscFixtures/Conformance/Stateful/NestedOperations/NestedOperations.cs
using SpecGate.Annotations;

namespace SpecGate.CtscFixtures.Conformance.Stateful.NestedOperations;

[SpecEvent("Account")]
public sealed class Account
{
    [SpecEvent("balance")]
    public int Balance { get; set; }

    [SpecSetup("transfer", Spec = "fixture.nested_operations")]
    [SpecSetup("withdraw", Spec = "fixture.nested_operations")]
    [SpecSetup("deposit", Spec = "fixture.nested_operations")]
    public static Account Make() => new() { Balance = 100 };

    [SpecOperation("transfer", Spec = "fixture.nested_operations")]
    public void Transfer([SpecInput("amount")] int amount)
    {
        Withdraw(amount);
        Deposit(amount);
        _ = ObserveBalance(Balance);
    }

    [SpecOperation("observe_balance", Spec = "fixture.nested_operations")]
    public static int ObserveBalance([SpecInput("balance")] int balance) => balance;

    [SpecOperation("withdraw", Spec = "fixture.nested_operations")]
    public void Withdraw([SpecInput("amount")] int amount) => Balance -= amount;

    [SpecOperation("deposit", Spec = "fixture.nested_operations")]
    public void Deposit([SpecInput("amount")] int amount) => Balance += amount;
}

File: test/rust/crates/specgate-ctsc-fixtures/src/conformance/stateful/nested_operations.rs
use specgate::{SpecEvent, spec_operation, spec_setup};

#[derive(Debug, PartialEq, Eq, SpecEvent)]
#[spec_component("fixture.nested_operations")]
pub struct Account {
    #[spec_event]
    pub balance: i32,
}

#[spec_setup("transfer", spec = "fixture.nested_operations")]
#[spec_setup("withdraw", spec = "fixture.nested_operations")]
#[spec_setup("deposit", spec = "fixture.nested_operations")]
pub fn make_account() -> Account {
    Account { balance: 100 }
}

impl Account {
    #[spec_operation("transfer", spec = "fixture.nested_operations")]
    pub fn transfer(&mut self, amount: i32) {
        self.withdraw(amount);
        self.deposit(amount);
        Self::observe_balance(self.balance);
    }

    #[spec_operation("observe_balance", spec = "fixture.nested_operations")]
    pub fn observe_balance(balance: i32) -> i32 {
        balance
    }

    #[spec_operation("withdraw", spec = "fixture.nested_operations")]
    pub fn withdraw(&mut self, amount: i32) {
        self.balance -= amount;
    }

    #[spec_operation("deposit", spec = "fixture.nested_operations")]
    pub fn deposit(&mut self, amount: i32) {
        self.balance += amount;
    }
}

#[test]
fn parent_operation_calls_two_annotated_child_operations() {
    let mut account = make_account();
    account.transfer(20);
    assert_eq!(account.balance, 100);
}
